// @volar/shared/out/path.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFileInDir = exports.getUriByPath = exports.normalizeUri = exports.normalizeFileName = exports.getPathOfUri = void 0;
const vscode_uri_1 = require("vscode-uri");
const upath = require("upath");
function getPathOfUri(uri) {
    return upath.toUnix(vscode_uri_1.URI.parse(uri).fsPath);
}
exports.getPathOfUri = getPathOfUri;
function normalizeFileName(fsPath) {
    return upath.toUnix(vscode_uri_1.URI.file(fsPath).fsPath);
}
exports.normalizeFileName = normalizeFileName;
function normalizeUri(uri) {
    return vscode_uri_1.URI.parse(uri).toString();
}
exports.normalizeUri = normalizeUri;
function getUriByPath(rootUri, path) {
    return vscode_uri_1.URI.file(path).with({
        scheme: rootUri.scheme,
        authority: rootUri.authority,
    }).toString();
}
exports.getUriByPath = getUriByPath;
function isFileInDir(fileName, dir) {
    const relative = upath.relative(dir, fileName);
    return !!relative && !relative.startsWith('..') && !upath.isAbsolute(relative);
}
exports.isFileInDir = isFileInDir;
