// @volar/shared/out/uriMap.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUriMap = exports._ = void 0;
const vscode_uri_1 = require("vscode-uri");
const path_1 = require("./path");
exports._ = require("vscode-uri");
function createUriMap(map = new Map()) {
    const uriToUriKeys = new Map();
    const pathToUriKeys = new WeakMap();
    return {
        clear,
        values,
        uriDelete,
        uriGet,
        uriHas,
        uriSet,
        pathDelete,
        pathGet,
        pathHas,
        pathSet,
    };
    function getUriByUri(uri) {
        if (!uriToUriKeys.has(uri))
            uriToUriKeys.set(uri, normalizeUri(uri).toLowerCase());
        return uriToUriKeys.get(uri);
    }
    function getUriByPath(rootUri, path) {
        let map = pathToUriKeys.get(rootUri);
        if (!map) {
            map = new Map();
            pathToUriKeys.set(rootUri, map);
        }
        if (!map.has(path)) {
            map.set(path, (0, path_1.getUriByPath)(rootUri, path).toLowerCase());
        }
        return map.get(path);
    }
    function clear() {
        return map.clear();
    }
    function values() {
        return map.values();
    }
    function uriDelete(_uri) {
        return map.delete(getUriByUri(_uri));
    }
    function uriGet(_uri) {
        return map.get(getUriByUri(_uri));
    }
    function uriHas(_uri) {
        return map.has(getUriByUri(_uri));
    }
    function uriSet(_uri, item) {
        return map.set(getUriByUri(_uri), item);
    }
    function pathDelete(rootUri, path) {
        return uriDelete(getUriByPath(rootUri, path));
    }
    function pathGet(rootUri, path) {
        return uriGet(getUriByPath(rootUri, path));
    }
    function pathHas(rootUri, path) {
        return uriGet(getUriByPath(rootUri, path));
    }
    function pathSet(rootUri, path, item) {
        return uriSet(getUriByPath(rootUri, path), item);
    }
}
exports.createUriMap = createUriMap;
function normalizeUri(uri) {
    try {
        return vscode_uri_1.URI.parse(uri).toString();
    }
    catch (_a) {
        return '';
    }
}
