// @volar/vue-language-service/out/utils/singleFileTypeScriptService.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSingleFileTypeScriptService = void 0;
const vscode_uri_1 = require("vscode-uri");
const shared = require("@volar/shared");
let projectVersion = 0;
let doc;
let scriptFileName;
let scriptSnapshot;
let service;
const host = {
    readFile: () => undefined,
    fileExists: fileName => fileName === scriptFileName,
    getProjectVersion: () => projectVersion.toString(),
    getScriptVersion: () => projectVersion.toString(),
    getCompilationSettings: () => ({ allowJs: true, jsx: 1 }),
    getScriptFileNames: () => [scriptFileName],
    getScriptSnapshot: fileName => {
        if (fileName === scriptFileName) {
            return scriptSnapshot;
        }
    },
    getCurrentDirectory: () => '',
    getDefaultLibFileName: () => '',
};
function getSingleFileTypeScriptService(ts, ts2, _doc, getConfiguration) {
    if (!service) {
        service = ts2.createLanguageService(ts, host, ts.createLanguageService(host), getConfiguration, vscode_uri_1.URI.file('/'));
    }
    projectVersion++;
    doc = _doc;
    scriptFileName = shared.getPathOfUri(_doc.uri);
    scriptSnapshot = ts.ScriptSnapshot.fromString(doc.getText());
    return service;
}
exports.getSingleFileTypeScriptService = getSingleFileTypeScriptService;
