// @volar/vue-language-server/out/requests.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.FsReadDirectoryRequest = exports.FsReadFileRequest = exports.FsStatRequest = exports.DetectDocumentNameCasesRequest = exports.ReloadProjectNotification = exports.WriteVirtualFilesNotification = exports.VerifyAllScriptsNotification = exports.AutoInsertRequest = exports.D3Request = exports.GetMatchTsConfigRequest = exports.FindFileReferenceRequest = exports.GetEditorSelectionRequest = exports.GetDocumentPrintWidthRequest = exports.GetDocumentNameCasesRequest = exports.ShowReferencesNotification = exports.GetDocumentContentRequest = void 0;
const vscode = require("vscode-languageserver-protocol");
/**
 * Client Requests
 */
var GetDocumentContentRequest;
(function (GetDocumentContentRequest) {
    GetDocumentContentRequest.type = new vscode.RequestType('vscode/content');
})(GetDocumentContentRequest = exports.GetDocumentContentRequest || (exports.GetDocumentContentRequest = {}));
var ShowReferencesNotification;
(function (ShowReferencesNotification) {
    ShowReferencesNotification.type = new vscode.NotificationType('vue.findReferences');
})(ShowReferencesNotification = exports.ShowReferencesNotification || (exports.ShowReferencesNotification = {}));
var GetDocumentNameCasesRequest;
(function (GetDocumentNameCasesRequest) {
    GetDocumentNameCasesRequest.type = new vscode.RequestType('volar/getAttrNameCaseClient');
})(GetDocumentNameCasesRequest = exports.GetDocumentNameCasesRequest || (exports.GetDocumentNameCasesRequest = {}));
var GetDocumentPrintWidthRequest;
(function (GetDocumentPrintWidthRequest) {
    GetDocumentPrintWidthRequest.type = new vscode.RequestType('vue/getDocumentWordWrapColumn');
})(GetDocumentPrintWidthRequest = exports.GetDocumentPrintWidthRequest || (exports.GetDocumentPrintWidthRequest = {}));
var GetEditorSelectionRequest;
(function (GetEditorSelectionRequest) {
    GetEditorSelectionRequest.type = new vscode.RequestType0('vue/activeSelection');
})(GetEditorSelectionRequest = exports.GetEditorSelectionRequest || (exports.GetEditorSelectionRequest = {}));
var FindFileReferenceRequest;
(function (FindFileReferenceRequest) {
    FindFileReferenceRequest.type = new vscode.RequestType('vue/findFileReference');
})(FindFileReferenceRequest = exports.FindFileReferenceRequest || (exports.FindFileReferenceRequest = {}));
/**
 * Server Requests
 */
var GetMatchTsConfigRequest;
(function (GetMatchTsConfigRequest) {
    GetMatchTsConfigRequest.type = new vscode.RequestType('volar/tsconfig');
})(GetMatchTsConfigRequest = exports.GetMatchTsConfigRequest || (exports.GetMatchTsConfigRequest = {}));
var D3Request;
(function (D3Request) {
    D3Request.type = new vscode.RequestType('volar/d3');
})(D3Request = exports.D3Request || (exports.D3Request = {}));
var AutoInsertRequest;
(function (AutoInsertRequest) {
    AutoInsertRequest.type = new vscode.RequestType('vue/autoInsert');
})(AutoInsertRequest = exports.AutoInsertRequest || (exports.AutoInsertRequest = {}));
var VerifyAllScriptsNotification;
(function (VerifyAllScriptsNotification) {
    VerifyAllScriptsNotification.type = new vscode.NotificationType('volar.action.verifyAllScripts');
})(VerifyAllScriptsNotification = exports.VerifyAllScriptsNotification || (exports.VerifyAllScriptsNotification = {}));
var WriteVirtualFilesNotification;
(function (WriteVirtualFilesNotification) {
    WriteVirtualFilesNotification.type = new vscode.NotificationType('volar.action.writeVirtualFiles');
})(WriteVirtualFilesNotification = exports.WriteVirtualFilesNotification || (exports.WriteVirtualFilesNotification = {}));
var ReloadProjectNotification;
(function (ReloadProjectNotification) {
    ReloadProjectNotification.type = new vscode.NotificationType('volar.action.reloadProject');
})(ReloadProjectNotification = exports.ReloadProjectNotification || (exports.ReloadProjectNotification = {}));
var DetectDocumentNameCasesRequest;
(function (DetectDocumentNameCasesRequest) {
    DetectDocumentNameCasesRequest.type = new vscode.RequestType('volar/getTagNameCaseServer');
})(DetectDocumentNameCasesRequest = exports.DetectDocumentNameCasesRequest || (exports.DetectDocumentNameCasesRequest = {}));
/**
 * FS request for browser
 */
var FsStatRequest;
(function (FsStatRequest) {
    FsStatRequest.type = new vscode.RequestType('fs/stat');
})(FsStatRequest = exports.FsStatRequest || (exports.FsStatRequest = {}));
var FsReadFileRequest;
(function (FsReadFileRequest) {
    FsReadFileRequest.type = new vscode.RequestType('fs/readFile');
})(FsReadFileRequest = exports.FsReadFileRequest || (exports.FsReadFileRequest = {}));
var FsReadDirectoryRequest;
(function (FsReadDirectoryRequest) {
    FsReadDirectoryRequest.type = new vscode.RequestType('fs/readDirectory');
})(FsReadDirectoryRequest = exports.FsReadDirectoryRequest || (exports.FsReadDirectoryRequest = {}));
